# Project3 Used Library  
***  
### **INDEX**  
>  
> 1. 디렉터리 다루기 - os.path, pathlib <br>
> 2. 파일 읽기, 저장 - fileinput, pickle <br>
> **3. 파일 찾기, 복사, 이동 - glob, fnmatch, shutil**<br>
> 4. 파일 압축 - zlib, gzip, zipfile, tarfile <br>

## 3. 파일 찾기, 복사, 이동

### glob

> - `glob` 는 패턴을 이용하여 파일을 검색할 때 사용하는 모듈

#### 파일 확장자로 찾기

In [17]:
import glob
import os

In [18]:
# 현재 디렉토리 확인
os.getcwd()

'c:\\Users\\ljg98\\FastC\\Fourty7\\basic project\\project3'

In [19]:
# txt 파일 찾기 - 현재 경로
for filename in glob.glob("*.txt"):
    print(filename)

00 폴더 관리 프로그램.txt
9999. fake_file.txt


In [20]:
# txt 파일 찾기 - 하위 경로
for filename in glob.glob("**/*.txt"):
    print(filename)

sample\list.txt
sample\새파일1.txt
sample\새파일2.txt
sample\새파일3.txt
sample_org\새파일1.txt
sample_org\새파일2.txt
sample_org\새파일3.txt


In [21]:
# txt 파일 찾기 - 현재와 하위 경로 모두 포함
for filename in glob.glob("**/*.txt", recursive=True):
    print(filename)

00 폴더 관리 프로그램.txt
9999. fake_file.txt
sample\list.txt
sample\새파일1.txt
sample\새파일2.txt
sample\새파일3.txt
sample_org\새파일1.txt
sample_org\새파일2.txt
sample_org\새파일3.txt


#### 파일명 글자수로 찾기
- ? 는 문자의 종류는 상관 X, 단 한개를 뜻할 때

In [31]:
# 글자수 4
for filename in glob.glob("**/????.*", recursive=True):
    print(filename)



9999. fake_file.txt
list.pkl
sample\새파일1.txt
sample\새파일2.txt
sample\새파일3.txt
sample_org\새파일1.txt
sample_org\새파일2.txt
sample_org\새파일3.txt


이 방식이 조금 문제가 있는게 <br>
만약  9999. fake_file.txt 와 같이 파일명에 ' . '(온점)이 들어가 있는 경우<br>
그 이하를 확장자 명으로 인식한다<br>

**파일명 저장할 때 . 지양!!**

#### 문자열 포함 파일명 찾기
- *은 문자 종류, 길이 상관 X

In [29]:
for filename in glob.glob("**/[a-z][a-z][a-z][a-z].*", recursive=True):
    print(filename)

list.pkl


In [32]:
#새파일 + anything
for filename in glob.glob("**/새파일*.*", recursive=True):
    print(filename)

sample\새파일1.txt
sample\새파일2.txt
sample\새파일3.txt
sample\새파일987654.txt
sample_org\새파일1.txt
sample_org\새파일2.txt
sample_org\새파일3.txt


In [33]:
# anything + project + anything

for filename in glob.glob("**/*project*.*", recursive=True):
    print(filename)

01  project3_outline.ipynb
02-1 project3_lib.ipynb
02-2 project3_lib.ipynb
02-3 project3_lib.ipynb
02-4 project3_lib.ipynb


### fnmatch

> - `fnmatch` 는 `glob` 와 동일하게 특정한 패턴을 따르는 파일명을 찾아주는 모듈입니다.
> - 파일명 매칭 여부를 `True`, `False` 형태로 반환하기 때문에 `os.listdir()` 함수와 함께 사용하는 것이 특징입니다

In [34]:
import fnmatch
import os

#### 탐색 조건
- 탐색범위는 sample 폴더
- 파일명은 '새'로 시작한다.
- 확장자는 텍스트 파일을 의미하는 .txt 이다.
- 확장자를 제외한 파일명의 길이는 4이다.
- 파일명의 마지막 문자는 숫자이다.

In [36]:
for filename in os.listdir('./sample'):
    if fnmatch.fnmatch(filename, '새??[0-9].txt'):
        print(filename)

새파일1.txt
새파일2.txt
새파일3.txt


### shutil

> - `shutil` 은 파일을 복사하거나 이동할때 사용하는 내장 모듈입니다.
> - `copy`, `copy2` : 파일 복사
> - `move` : 파일 이동, 확장자 변경

In [37]:
import shutil

In [39]:
# 파일 복사하기
shutil.copy("./sample/새파일1.txt", "./sample/새파일1_복사본.txt")

'./sample/새파일1_복사본.txt'

In [40]:
# 메타정보 복사
shutil.copy2("./sample/새파일1.txt", "./sample/새파일1_복사본_메타.txt")

'./sample/새파일1_복사본_메타.txt'

In [41]:
# 확장자 바꾸기
shutil.move("./sample/새파일1_복사본.txt", "./sample/새파일1_복사본.py")

'./sample/새파일1_복사본.py'

In [42]:
# 확장자 바꾸기
shutil.move("./sample/새파일1_복사본.py", "./sample/새파일1_복사본.txt")

'./sample/새파일1_복사본.txt'

In [43]:
# 파일 이동하기
shutil.move("./sample/새파일1_복사본.txt", "./sample_copy/새파일1_복사본.txt")
shutil.move("./sample/새파일1_복사본_메타.txt", "./sample_copy/새파일1_복사본_메타.txt")

'./sample_copy/새파일1_복사본_메타.txt'

#### 메타 정보란?
- by chatGPT

메타 정보(meta-information)는 데이터에 관한 데이터로, 주어진 데이터의 특정 측면에 대한 정보를 나타냅니다. <br>
이는 데이터의 구조, 형식, 속성 등에 대한 정보를 포함할 수 있습니다.<br>
예를 들어, 이미지 파일의 메타 정보에는 이미지의 크기, 해상도, 색상 프로파일, 촬영일시 등의 정보가 포함될 수 있습니다.<br>
웹사이트에서의 메타 정보는 웹페이지의 제목, 설명, 키워드 등을 포함하며, 이는 검색 엔진에서 웹페이지를 색인하고 검색 결과에 표시하는 데 사용됩니다.